Add Nerves lesson (#1766)
elixirschool/elixirschool: en/lessons/advanced/nerves.md

---
version: 1.0.0
title: Nerves
---

## Introduction and requirements

We will be talking about Nerves in this lesson. Nerves project is a framework for using Elixir in embedded software development. As their site says, `Craft and deploy bulletproof embedded software in Elixir`. Due to the topic, this lesson will be a bit different from other Elixir School lessons. Nerves is an advanced topic and require some advanced system setup and additional hardware, as well as understanding what you are doing. You've been warned!

> Today we are going to be exploring this land of embedded development. So, if you have hardware from their list of [supported targets](https://hexdocs.pm/nerves/targets.html) , card reader with a memory card supported by the hardware of your choice + wired networking connection to access this device by the network, we are ready to start. However, we suggest you have a Raspberry PI, due to it having controllable LED onboard. Otherwise, you will need to build a scheme similar to [this one](https://github.com/fhunleth/elixir_ale#gpio) to make use of external LED. We also suggest you have a screen connected to your hardware of choice - this will simplify configuration a lot.

Reviewer: PI -> pi.

It'd be kind of a pain, but for the sake of diffing some of us have been writing in a style of one-sentence-per-line for very friendly diffing as time goes on and edits come in. Might not be worth doing here, but figured I'd note.

Author: Thanks, fixed. 

Hm, I'll look into other lessons.

Author: I've checked, and more complex lessons like https://elixirschool.com/en/lessons/advanced/typespec/ and https://elixirschool.com/en/lessons/advanced/gen-stage/ use multiple sentences per paragraph due to information density. I guess this is fine here.

If you have a screen connected - you should see a Linux boot sequence on it after powering up the device with this memory card inserted.

## Setting up networking

Next step - if we want a networked led device - is setting up a network. Nerves ecosystem has some exciting packages, and [nerves_network](https://github.com/nerves-project/nerves_network) is one of them.

It's already present in your project - as a dependency of `nerves_init_gadget`. However, by default, it uses DHCP (see config for it in `config/config.exs` after the `config :nerves_init_gadget`), but in case of this tutorial, it's preferred to have a static IP address.

To set static networking, you need to add the configuration to `config/config.exs`:

```
# Statically assign an address
config :nerves_network, :default,
  eth0: [
    ipv4_address_method: :static,
    ipv4_address: "192.168.88.2",
    ipv4_subnet_mask: "255.255.255.0",
    nameservers: ["8.8.8.8", "8.8.4.4"]
  ]
 ```

Please note that this config is for a wired connection. If you want to use wireless connection - take a look at [Nerves network documentation](https://github.com/nerves-project/nerves_network#wifi-networking).

> Note that you need to use your local network parameters here - in my network there is an unallocated IP`192.168.88.2`, which I am going to use. However, in your case, it may differ.

Reviewer 2: ```suggestion
Note that you need to use your local network parameters here - in my network there is an unallocated IP `192.168.88.2`, which I am going to use. However, in your case, it may differ.
```

After adding this, let's try burning and running the changed version of the firmware via `mix firmware.burn`.

If you did as I and started the `ping 192.168.88.2` command before powering the device up, you will be able to see something like this:

```
Request timeout for icmp_seq 206
Request timeout for icmp_seq 207
64 bytes from 192.168.88.2: icmp_seq=208 ttl=64 time=2.247 ms
64 bytes from 192.168.88.2: icmp_seq=209 ttl=64 time=2.658 ms
```

This output means that the device now is reachable from the network!

> Seems like the best time to launch a new crowdfunded hardware startup, isn't it?

Author: @doomspork What do you think, is this OK thing to have in Elixir School lesson?

> Seems like the best time to launch a new crowdfunded hardware startup, isn't it?

Reviewer 3: @brain-geek I think we should remove this. I went through the lesson and I concur with @tommorris, I'm not sure this fits the tone of Elixir School.

> Seems like the best time to launch a new crowdfunded hardware startup, isn't it?

Author: Yeah, we usually don't use humor in these lessons, therefor I was not sure. Thanks for answering, fixed.

## Network firmware burning

Previously, we were burning SD cards and then physically pushed them into hardware. While it is okay for starters, it's more straightforward to have a network update. That's what `nerves_firmware_ssh` does. It is already present in your project by default and configured to auto-detect and find ssh keys in your folder.

But to use network firmware update, you need to generate an upload script via  `mix firmware.gen.script`. This command will generate new `upload.sh` script for updating the firmware.

If the network has been functional after the previous step, that's it!

You can finally stop switching SD cards: `mix firmware && ./upload.sh 192.168.88.2` is a combination of two commands: first creates a firmware, and the second one burns it via the network and reboots the device.

> _Hint: `ssh 192.168.88.2` gives you a proper IEx shell on the device in the context of the app. _

Author: My idea was that 'improper' would be the one without any application environment, or not running in context of device... But yeah, this change makes a lot of sense.

_Troubleshooting: If you don't have an existing ssh key in your home folder, you will have an error `No SSH public keys found in ~/.ssh.`. In this case, you will need to run `ssh-keygen` and re-burn the firmware to use the network update feature._

## Setting up the LED control

To interact with LEDs, you need [nerves_leds](https://github.com/nerves-project/nerves_leds) package installed which, as you know, is done by adding `{:nerves_leds, "~> 0.8", targets: @all_targets},` to `mix.exs` file.

After setting up the dependency, you need to configure the LED list for the given device. For example, for all Raspberry Pi models, there is only one LED onboard: `led0`. Let's use it by adding a `config :nerves_leds, names: [green: "led0"]` line to the `config/config.exs`.

For other devices, you can take a look at the [corresponding part of the nerves_examples project](https://github.com/nerves-project/nerves_examples/tree/master/hello_leds/config).

> After configuring the LED itself, we surely need to control it somehow. To do that, we will add a GenServer in `lib/network_led/blinker.ex` with these contents:

Reviewer 4: May be worth pointing to the [GenServer](https://elixirschool.com/en/lessons/advanced/otp-concurrency/#genserver) section of Elixir School site. Elixir devs are going to hopefully know what a GenServer is, but those from an embedded development background who are new to Elixir may not.

```elixir
defmodule NetworkLed.Blinker do
  use GenServer

  @moduledoc """
    Simple GenServer to control GPIO #18.
  """

  require Logger
  alias Nerves.Leds

  def start_link(state \\ []) do
    GenServer.start_link(__MODULE__, state, name: __MODULE__)
  end

  def init(state) do
    enable()

    {:ok, state}
  end

  def handle_cast(:enable, state) do
    Logger.info("Enabling LED")
    Leds.set(green: true)

    {:noreply, state}
  end

  def handle_cast(:disable, state) do
    Logger.info("Disabling LED")
    Leds.set(green: false)

    {:noreply, state}
  end

  def enable() do
    GenServer.cast(__MODULE__, :enable)
  end

  def disable() do
    GenServer.cast(__MODULE__, :disable)
  end
end

```

To enable this, you also need to add it to the supervision tree in `lib/network_led/application.ex`: add `{NetworkLed.Blinker, name: NetworkLed.Blinker}` under the `  def children(_target) do` group.

Notice that Nerves has two different supervision trees in application - one for the host machine and one for actual devices.

> After this - that's it! You actually can upload the firmware and check that `NetworkLed.Blinker.disable()` disables LED (which is enabled by default in code), and `NetworkLed.Blinker.enable()` enables it.

Reviewer 4: How do I check? IEx on the host, or on the target device? Worth just noting.

Author: Thanks for noting, added.

> After this - that's it! You actually can upload the firmware and check that `NetworkLed.Blinker.disable()` disables LED (which is enabled by default in code), and `NetworkLed.Blinker.enable()` enables it.

Reviewer 4: ```suggestion
After this - that's it! You actually can upload the firmware and check that `NetworkLed.Blinker.disable()` turns the LED off (which is enabled by default in code), and `NetworkLed.Blinker.enable()` turns it on.
```

We have control over the LED from the command prompt!

> Now the only missing piece of the puzzle we need is controlling it via the web interface.

Reviewer 4: ```suggestion
Now the only missing piece of the puzzle left is to control the LED via the web interface.
```

---
version: 1.0.0
title: Nerves
---

## Introduction and requirements

We will be talking about Nerves in this lesson. Nerves project is a framework for using Elixir in embedded software development. As their site says, `Craft and deploy bulletproof embedded software in Elixir`. Due to the topic, this lesson will be a bit different from other Elixir School lessons. Nerves is an advanced topic and require some advanced system setup and additional hardware, as well as understanding what you are doing. 

Today we are going to be exploring this land of embedded development. So, if you have hardware from their list of [supported targets](https://hexdocs.pm/nerves/targets.html), card reader with a memory card supported by the hardware of your choice + wired networking connection to access this device by the network, we are ready to start. 

However, we suggest you have a Raspberry Pi, due to it having controllable LED onboard. We also advise you have a screen connected to your hardware of choice - this will simplify debugging a lot due to having IEx.

## Setup

> Nerves project itself has an excellent [Getting started guide](https://hexdocs.pm/nerves/getting-started.html), but let's try approaching from a "fewer words, more code" approach.

Reviewer 5: can you reword this? because you have two approaches in there :) something like `let's try viewing this from a "fewer words, more code" perspective`.

> Firstly, you need an environment set up. You can find the guide in the [Installation](https://hexdocs.pm/nerves/installation.html) part of Nerves wiki. You may want to skip it, however, please make sure that you have the same version of both OTP and Elixir they are mentioning. This check could save you sometime later on. At the moment of writing, any OTP 21 Elixir should work.

Reviewer 5: some time

any OTP 21 Elixir (version | subversion) should work

Author: Not sure what you mean in the second part, rephrased.

After running all the setup part, you should be ready to build your first Nerves project! 

> We are going to develop a standard for embedded development "Hello world" project - blinking led controlled by web API. 

Reviewer 5: a blinking led?

## Creating a project

To generate a new project, run `mix nerves.new network_led` and answer `Y` when prompted whether to fetch and install dependencies. 

> As a result, you should have such output:

Reviewer 5: a similar output

```
Your Nerves project was created successfully.

You should now pick a target. See https://hexdocs.pm/nerves/targets.html#content
for supported targets. If your target is on the list, set `MIX_TARGET`
to its tag name:

For example, for the Raspberry Pi 3 you can either
  $ export MIX_TARGET=rpi3
Or prefix `mix` commands like the following:
  $ MIX_TARGET=rpi3 mix firmware

If you will be using a custom system, update the `mix.exs`
dependencies to point to desired system's package.

Now download the dependencies and build a firmware archive:
  $ cd network_led
  $ mix deps.get
  $ mix firmware

If your target boots up using an SDCard (like the Raspberry Pi 3),
then insert an SDCard into a reader on your computer and run:
  $ mix firmware.burn

Plug the SDCard into the target and power it up. See target documentation
above for more information and other targets.
```

Our project has been generated and is ready to be flashed to our test device! Let's try it now! 

> In my case, the device is Raspberry 3, so I chose `MIX_TARGET=rpi3`. In your case, it may be different. However, all the other commands will remain the same in this part. 

Reviewer 5: the devise is a Raspberry Pi 3

> Let's set up dependencies first:

Reviewer 5: our dependencies

If you have a screen connected - you should see a Linux boot sequence on it after powering up the device with this memory card inserted.

## Setting up networking

Next step - if we want a networked led device - is setting up a network. Nerves ecosystem has some exciting packages, and [nerves_network](https://github.com/nerves-project/nerves_network) is one of them. 

It's already present in your project - as a dependency of `nerves_init_gadget`. However, by default, it uses DHCP (see config for it in `config/config.exs` after the `config :nerves_init_gadget`), but in case of this tutorial, it's preferred to have a static IP address. 

To set static networking, you need to add the configuration to `config/config.exs`:

```
# Statically assign an address
config :nerves_network, :default,
  eth0: [
    ipv4_address_method: :static,
    ipv4_address: "192.168.88.2", 
    ipv4_subnet_mask: "255.255.255.0",
    nameservers: ["8.8.8.8", "8.8.4.4"]
  ]
 ```

Please note that this config is for a wired connection. If you want to use wireless connection - take a look at [Nerves network documentation](https://github.com/nerves-project/nerves_network#wifi-networking).

> In my network, I will be going the device with IP `192.168.88.2` in subnet `255.255.255.0`. However, in your case, it may differ.

Reviewer 5: can you reword this phrase? It reads a bit off.

Author: Coud you check out once again? Rephrased.

> After adding this let's try burning and running the changed version of the firmware via `mix firmware.burn`. 

Reviewer 5: Afterr adding this, let's

If you did as I and started the `ping 192.168.88.2` command before powering the device up, you will be able to see something like this:

```
Request timeout for icmp_seq 206
Request timeout for icmp_seq 207
64 bytes from 192.168.88.2: icmp_seq=208 ttl=64 time=2.247 ms
64 bytes from 192.168.88.2: icmp_seq=209 ttl=64 time=2.658 ms
```

This output means that the device now is reachable from the network! 

Seems like the best time to launch a new crowdfunded hardware startup, isn't it?

## Network firmware burning

Previously, we were burning SD cards and then physically pushed them into hardware. While it is okay for starters, it's more straightforward to have a network update. That's what `nerves_firmware_ssh` does. It is already present in your project by default and configured to auto-detect and find ssh keys in your folder. 

But to use network firmware update, you need to generate an upload script via  `mix firmware.gen.script`. This command will generate new `upload.sh` script for updating the firmware.

If the network has been functional after the previous step, that's it! 

> You can finally stop switching SD cards: `mix firmware&&./upload.sh 192.168.88.2` automatically creates a firmware, burns it via the network and reboots the device. 

Reviewer 5: add some spaces in the command

_Hint: `ssh 192.168.88.2` gives you a proper IEx shell on the device in the context of the app. _

_Troubleshooting: If you don't have an existing ssh key in your home folder, you will have an error `No SSH public keys found in ~/.ssh.`. In this case, you will need to run `ssh-keygen` and re-burn the firmware to use the network update feature._

## Setting up the LED control

To interact with LEDs, you need [nerves_leds](https://github.com/nerves-project/nerves_leds) package installed which, as you know, is done by adding `{:nerves_leds, "~> 0.8", targets: @all_targets},` to `mix.exs` file. 

After setting up the dependency, you need to configure the LED list for the given device. For example, for all Raspberry Pi models, there is only one LED onboard: `led0`. Let's use it by adding a `config :nerves_leds, names: [green: "led0"]` line to the `config/config.exs`. 

For other devices, you can take a look at the [corresponding part of the nerves_examples project](https://github.com/nerves-project/nerves_examples/tree/master/hello_leds/config).

> After configuring the LEDs itself, we surely need to control them somehow. To do that, we will add a GenServer in `lib/network_led/blinker.ex` with these contents:

Reviewer 5: After configuring the LEDs themselves, we surely need to control them somehow. Or in the singular form for one LED.

```elixir
defmodule NetworkLed.Blinker do
  use GenServer

  @moduledoc """
    Simple GenServer to control GPIO #18.
  """

  require Logger
  alias Nerves.Leds

  def start_link(state \\ []) do
    GenServer.start_link(__MODULE__, state, name: __MODULE__)
  end

  def init(state) do
    enable()

    {:ok, state}
  end

  def handle_cast(:enable, state) do
    Logger.info("Enabling LED")
    Leds.set(green: true)

    {:noreply, state}
  end

  def handle_cast(:disable, state) do
    Logger.info("Disabling LED")
    Leds.set(green: false)

    {:noreply, state}
  end

  def enable() do
    GenServer.cast(__MODULE__, :enable)
  end

  def disable() do
    GenServer.cast(__MODULE__, :disable)
  end
end

```

> And to enable it you need to also add it to supervision tree in `lib/network_led/application.ex`: add `{NetworkLed.Blinker, name: NetworkLed.Blinker}` under the `  def children(_target) do` group. 

Reviewer 5: To enable this, you also need to add it to supervision tree

First, we will add `{:plug_cowboy, "~> 2.0"},` to `mix.exs` and install the dependencies. 

Then, add the actual process to process those requests in `lib/network_led/http.ex` :

```elixir
defmodule NetworkLed.Http do
  use Plug.Router

  plug(:match)
  plug(:dispatch)

  get("/", do: send_resp(conn, 200, "Feel free to use API endpoints!"))

  get "/enable" do
    NetworkLed.Blinker.enable()
    send_resp(conn, 200, "LED enabled")
  end

  get "/disable" do
    NetworkLed.Blinker.disable()
    send_resp(conn, 200, "LED disabled")
  end

  match(_, do: send_resp(conn, 404, "Oops!"))
end
```

And, the final step - add `{Plug.Cowboy, scheme: :http, plug: NetworkLed.Http, options: [port: 80]}` to the application supervision tree.

After the firmware update, you can try it! `http://192.168.88.2/` is returning plain text response, and `http://192.168.88.2/enable` with `http://192.168.88.2/disable` disable and enable that LED!

You can even pack Phoenix-powered user interfaces into your Nerves app, however, it [will require some tweaking](https://github.com/nerves-project/nerves/blob/master/docs/User%20Interfaces.md#phoenix-web-interfaces). 

> 

Reviewer 5: remove trailing space.

> 

Reviewer 5: Thanks @brain-geek!

All in all very quick and informative, unfortunately I cannot test it in my Rpi3 as my sd card reader has left for a better place. I am going to buy a new one though and check again, but it seems solid as a starting lesson.

I have left some minor corrections for you.

Author: Thanks a lot @gemantzu , applied all your suggestions!
